controlpanel/api/tools.py

```diff
@@ -0,0 +1,237 @@
+from collections import UserDict
+import secrets
+
+from django.conf import settings
+from django.utils.functional import cached_property
+from rest_framework.exceptions import APIException
+
+from controlpanel.api.helm import Helm
+from controlpanel.kubeapi.views import client
+
+
+IDLED = "mojanalytics.xyz/idled"
+
+SUPPORTED_TOOL_NAMES = list(settings.TOOLS.keys())
+
+
+class DuplicateTool(Exception):
+    pass
+
+
+class ToolNotDeployed(APIException):
+    status_code = 400
+    default_detail = f"Specified tool not deployed"
+    default_code = "tool_not_deployed"
+
+
+class UnsupportedToolException(APIException):
+    status_code = 400
+    supported = ", ".join(SUPPORTED_TOOL_NAMES)
+    default_detail = f"Unsupported tool, must be one of: {supported}"
+    default_code = "unsupported_tool"
+
+
+class ToolRepository(UserDict):
+
+    def __getitem__(self, key):
+        if key not in self.tools:
+            raise UnsupportedToolException(key)
+
+        if key not in self.tools:
+            raise UnsupportedToolException(key)
+
+        return super().__getitem__(key)
+
+    def __setitem__(self, key, value):
+        if key not in SUPPORTED_TOOL_NAMES:
+            raise UnsupportedToolException(key)
+
+        if key in self.tools:
+            raise DuplicateTool(key)
+
+        super().__setitem__(key, value)
+
+
+class ToolMeta(type):
+    """Adds Tool subclasses to repository"""
+
+    # repository is shared across all Tool subclasses
+    _repository = ToolRepository()
+
+    def __init__(cls, name, bases, clsdict):
+        super().__init__(name, bases, clsdict)
+        if 'name' not in clsdict:
+            raise AttributeError(f"{name}.name")
+        if name != "Tool":
+            cls._repository[clsdict['name']] = cls
```

Comment:
#### 1e. Refactor Tools code
> Lines 55 - 66

This metaclass automatically adds subclasses of `Tool` to a shared `ToolRepository` instance. This removes (hides) the module-level `Tools` variable and encapsulates it to avoid it being messed with. Also, we no longer have to remember to add new `Tool` subclasses to the repository manually.

[Next →](https://github.com/ministryofjustice/analytics-platform-control-panel/pull/610/files#diff-aad7b41260663fb400ff942a02fd3d14R69)

Reply:
It's only a little one. But I think it's a good idea for the reasons I've given. What are your reasons against it?